zap/src/main/java/org/parosproxy/paros/network/HttpMessage.java

```diff
@@ -495,11 +495,23 @@ public void setRequestBody(String body) {
      */
     public void setRequestBody(byte[] body) {
         getRequestBody().setBody(body);
-        getRequestBody().setCharset(getRequestHeader().getCharset());
+        setBodyCharset(getRequestBody(), getRequestHeader());
 
         setContentEncodings(getRequestHeader(), getRequestBody());
     }
 
+    private static void setBodyCharset(HttpBody body, HttpHeader header) {
+        String charset = header.getCharset();
+        body.setCharset(charset);
+
+        if (charset != null && !charset.equalsIgnoreCase(body.getCharset())) {
+            LOGGER.warn(
+                    "Failed to set charset {} from content-type value: {}",
+                    charset,
+                    header.getNormalisedContentTypeValue());
```

Comment:
Is it worth including the raw value as well?

Reply:
The normalised value is just lower cased, the charset is case insensitive.